src/snarks/groth16/powersoftau_utils.cpp

```diff
@@ -8,6 +10,16 @@ namespace
 
 // Utility functions
 
+template<typename Fr, typename Gr>
+static void basic_radix2_iFFT(std::vector<Gr> &as, const Fr &omega_inv)
```

Comment:
> Do you really do an inverse FFT (iFFT) here?

I guess one could argue it's not exactly iFFT because of the `omega_inv` parameter, but it's iFFT in the sense that if we call `FFT(., omega_inv)` we get back the original powers.  But if you prefer to change the name then just let me know.

> I'd mention Section 3 of ...

Good point.  I added a comment in the code.

Reply:
> I guess one could argue it's not exactly iFFT because of the omega_inv parameter, but it's iFFT in the sense that if we call FFT(., omega_inv) we get back the original powers. But if you prefer to change the name then just let me know.

Ah, you passed `omega_inv` to directly retrieve the `P(omega_inv^i)`?
I had the `Claim 3.2` and the `subsection 3.3` of https://eprint.iacr.org/2017/602.pdf in mind here. So you haven't leveraged the fact that you can retrieve the `P(omega_inv^i)`'s by just reversing the `P(omega^i)`'s, right?

I guess the reason I'm confused is because the result of your `basic_radix2_iFFT` function is a set of evaluated lagrange polynomials at `tau`. Thus, we are still dealing with evaluated polynomials, and not polynomials in coefficient form (to me iFFT refers to interpolation and thus refers to the transform from evaluation point form back to polynomials in coefficient form).

Reply:
> [...] So you haven't leveraged the fact that you can retrieve the P(omega_inv^i)'s by just taking the P(omega^i)'s in reverse order, right?

That's right.  I've used the equality $FFT(., omega_inv) / c_d = iFFT(., omega)$, which I think is what the rust implementation of powersoftau was doing i.e. it was just calling iFFT on the evaluation domain.  (Note I think there is a missing factor $c_d$ in that claim 3.2, and having to divide by that factor may be kind of suggestive of iFFT).

But I see the point (if I understood correctly).  The input to iFFT here is the powers, rather than the actual evaluations of some polynomial, and the output is not used as polynomial coefficients.  Since this function is static and only used for this trick, semantically "iFFT" may not be the best name.

That makes sense.  I'll rename it.  (Please let me know if I've missed the point :) ).

Reply:
>  (Note I think there is a missing factor $c_d$ in that claim 3.2, and having to divide by that factor may be kind of suggestive of iFFT).

Yes I spotted it also. The claim is wrong size we should multiply by `c_d`, but they didn't forget to mention to multiply by `1/d` to obtain the LAG_tau vector:
```
After getting the result from the FFT, we reverse the order of the vector and multiply each element by the scalar 1/d.
```

> The input to iFFT here is the powers, rather than the actual evaluations of some polynomial, and the output is not used as polynomial coefficients

Yes exact!

Reply:
Maybe something like `get_lagrange_from_powers` and a comment referring to the paper may be enough to clarify the purpose of the function.

Reply:
Yes, that's the kind of thing I had in mind.  Renamed (and moved the paper reference) in the latest.

Reply:
Thanks!